basket/base/rq.py

```diff
@@ -0,0 +1,231 @@
+# Exceptions we allow to retry, all others will abort retries.
+EXCEPTIONS_ALLOW_RETRY = [
+    IOError,
+    NewsletterException,
+    requests.RequestException,
+    RetryTask,
+    SilverpopResponseException,
+]
+
+
+# Our cached Redis connection.
+_REDIS_CONN = None
+
+
+def get_redis_connection(url=None, force=False):
+    """
+    Get a Redis connection.
+
+    Expects a URL including the db, or defaults to `os.environ["RQ_URL"]`.
+
+    Call example:
+        get_redis_connection("redis://localhost:6379/0")
+
+    """
+    global _REDIS_CONN
+
+    if url is None:
+        url = os.environ.get("RQ_URL")
+
+    if force or _REDIS_CONN is None:
+        _REDIS_CONN = redis.Redis.from_url(url)
+    return _REDIS_CONN
+
+
+def get_queue(queue="default"):
+    """
+    Get an RQ queue with our chosen parameters.
+
+    """
+    return Queue(
+        queue,
+        connection=get_redis_connection(),
+        is_async=settings.RQ_IS_ASYNC,
+        serializer=JSONSerializer,
+    )
+
+
+def get_worker(queues=None):
+    """
+    Get an RQ worker with our chosen parameters.
+
+    """
+    if queues is None:
+        queues = ["default"]
+
+    return SimpleWorker(
+        queues,
+        connection=get_redis_connection(),
+        disable_default_exception_handler=True,
+        exception_handlers=[store_task_exception_handler],
+        serializer=JSONSerializer,
+    )
+
+
+def enqueue_kwargs(func):
+    if isinstance(func, str):
+        task_name = func
+    else:
+        task_name = f"{func.__module__}.{func.__qualname__}"
+
+    # Start time is used to calculate the total time taken by the task, which includes queue time plus execution time of the task itself.
+    meta = {
+        "task_name": task_name,
+        "start_time": time(),
+    }
+
+    if settings.RQ_MAX_RETRIES == 0:
+        retry = None
+    else:
+        retry = Retry(settings.RQ_MAX_RETRIES, rq_exponential_backoff())
+
+    return {
+        "meta": meta,
+        "retry": retry,
+        "result_ttl": settings.RQ_RESULT_TTL,
+        "on_success": Callback(rq_on_success),
+        "on_failure": Callback(rq_on_failure),
+    }
+
+
+def rq_exponential_backoff():
+    """
+    Return an array of retry delays for RQ using an exponential back-off, using
+    jitter to even out the spikes, waiting at least 1 minute between retries.
+    """
+    if settings.DEBUG:
+        # While debugging locally, enable faster retries.
+        return [5 for n in range(settings.RQ_MAX_RETRIES)]
+    else:
+        return [max(60, random.randrange(min(settings.RQ_MAX_RETRY_DELAY, 2 * (2**n)))) for n in range(settings.RQ_MAX_RETRIES)]
+
+
+def log_timing(job):
+    if start_time := job.meta.get("start_time"):
+        total_time = int(time() - start_time) * 1000
+        statsd.timing(f"{job.meta['task_name']}.duration", total_time)
+
+
+def rq_on_success(job, connection, result, *args, **kwargs):
+    # Don't fire statsd metrics in maintenance mode.
```

Comment:
why not?

Reply:
In maintenance mode we're going to queue the task in the db rather than do the task, so we'd get timings for doing the db insert rather than timings for the task, which would confuse the timings. And a "success" for the db insert isn't the same as a "success" for the task.